lib/geo/wkb/encoder.ex

```diff
@@ -18,238 +36,169 @@ defmodule Geo.WKB.Encoder do
-        {nil, acc}
-      end)
-
-    writer
-  end
-
-  defp encode_coordinates(writer, %PolygonZ{coordinates: coordinates}) do
-    number_of_lines = Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8)
-    writer = Writer.write(writer, number_of_lines)
-
-    {_nils, writer} =
-      Enum.map_reduce(coordinates, writer, fn line, acc ->
-        acc = encode_coordinates(acc, %LineStringZ{coordinates: line})
-        {nil, acc}
-      end)
-
-    writer
-  end
-
-  defp encode_coordinates(writer, %MultiPoint{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%Point{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPointZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%PointZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiLineString{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%LineString{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiLineStringZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%LineStringZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPolygon{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%Polygon{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
-  end
-
-  defp encode_coordinates(writer, %MultiPolygonZ{coordinates: coordinates}) do
-    writer = Writer.write(writer, Integer.to_string(length(coordinates), 16) |> Utils.pad_left(8))
-
-    geoms =
-      Enum.map(coordinates, fn geom ->
-        encode!(%PolygonZ{coordinates: geom}, writer.endian)
-      end)
-      |> Enum.join()
-
-    Writer.write_no_endian(writer, geoms)
+  defp add_srid(type), do: type + @wkbsridflag
+
+  def encode!(geom, endian \\ :ndr)
+
+  for {endian, endian_atom, modifier} <- [{1, :ndr, quote(do: little)}, {0, :xdr, quote(do: big)}] do
```

Comment:
I think this is nice, seems to strike balance between convenience and not too magical.

You could even extend it to the following:

```elixir
  for {endian, byte, u32, f64} <- [
        {:ndr, 0, quote(do: 32), quote(do: float - 64)},
        {:xdr, 1, quote(do: little - 32), quote(do: little - float - 64)}
      ] do
   # `unquote(modifier)-32` becomes `unquote(u32)`
   # `unquote(modifier)-float-64` becomes `unquote(f64)`
end
```

but not sure if it this wouldn't be a bit too clever.

Reply:
I like that! I think it's the perfect amount of clever!

Reply:
Looks like it might  be a little harder to get that to work so I'll leave it be for now